docs/reference/remap2/features/type_safety.cue

```diff
@@ -0,0 +1,74 @@
+remap2: features: type_safety: {
+	title: "Type safety"
+	description: """
+		VRL implements progressive type safety, erroring at compilation-time if a type mismatch is detected. This
+		largely contributes to VRL's safety principle, ensuring scripts work as expected post-compilation.
+		"""
+
+	characteristics: {
+		errors: {
+			title: "Errors"
+			description: """
+				If VRL detects a type mismatch, it will produce a user-friendly compile-time error. For example, given
+				this VRL script:
+
+				```vrl
+				foo, err = 5
+				```
+
+				VRL will produce the following compile-time error:
+
+				```
+				error: unneeded error assignment
+				  ┌─ :2:1
+				  │
+				2 │ foo, err = 5;
+				  │ ^^^^^^^^   - because this expression cannot fail
+				  │ │
+				  │ this error assignment is unneeded
+				  │
+				  = hint: assign to "foo", without assigning to "err"
+				  = see language documentation at: https://vector.dev
+				```
+				"""
+		}
+		progressive: {
+			title: "Progressivenes"
+			description: """
+				VRL's type safety is _progressive_, meaning it will implement type safety for any value for which it
+				knows the type. Because obervability data can be quite unpredictable, it's not always known which
+				type a field might be, hence the _progressive_ nature of VRL's type safety. As VRL scripts are
+				evaluated, type information is built up and used at compile-time to enforce type safety. Let's look
+				at an example:
+
+				```vrl
+				.foo # any
+				.foo, err = downcase(.foo) # string
+				.foo = upcase(.foo) # string
```

Comment:
Actually, because `downcase` can fail, `.foo` is either a `string` or `null`, so this won't work as expected.

You could either write:

```rust
.foo # any
.foo, err = downcase(.foo) # string or null
.foo = upcase(to_string(.foo)) # string
```

Or use error coalescing in this case:

```rust
.foo # any
.foo = downcase(.foo) ?? "" # string
.foo = upcase(.foo) # string
```

Here's another solution that doesn't currently work in VRL, but should:

```rust
.foo # any
.foo, err = downcase(.foo) # string or null
.foo = if .foo != null { upcase(.foo) } # string
```